### PR TITLE
[openCL] add group param to conv for OpenCL backend

### PR DESCRIPTION
**Pull request by beicy**
OpenCL fast kernel for group (group > 1) convolution has some issues which are need to be investigated further. Now for group conv, we only call the slow kernel.

### CODE REVIEW

File: lib/Backends/OpenCL/OpenCL.cpp [hunk 1]
```diff
@@ -525,9 +525,9 @@ void OpenCLFunction::executeConvolution(const OCLConvolutionInst *CC) {
 
   // The global work for the N dimension (which covers all pixels) should be big
   // enough to cover all outputs for the correctness.
-  if (global[0] * fw_div_N < odim.h * odim.w) {
-    global[0] = (odim.h * odim.w - 1) / fw_div_N + 1;
-  }
+  //  if (global[0] * fw_div_N < odim.h * odim.w) {
+  //  global[0] = (odim.h * odim.w - 1) / fw_div_N + 1;
+  //}
```
Comment (opti-mix) [inline lib/Backends/OpenCL/OpenCL.cpp]: Can you add this fix in a separate PR or at least in a separate commit? It is not related to group convolutions at all.

File: lib/Backends/OpenCL/Transforms.cpp [hunk 2]
```diff
@@ -36,6 +36,11 @@ bool OCLBackend::transformPostLowering(Function *F,
   bool changed = false;
   for (auto &node : F->getNodes()) {
     if (auto *CN = dyn_cast<ConvolutionNode>(&node)) {
+      // TODO: OpenCL fast kernel itself has some issue with group > 1,
```
Comment (opti-mix) [inline lib/Backends/OpenCL/Transforms.cpp]: fast kernel -> fast convolution kernel

File: lib/Backends/OpenCL/OpenCL.cpp (hunk 1 above)
Comment (beicy) [reply lib/Backends/OpenCL/OpenCL.cpp]: OK. Will revert it in this PR and create a new one.

File: lib/Backends/OpenCL/OpenCL.h [hunk 3]
```diff
@@ -205,6 +205,13 @@ class OCLBackend final : public Backend {
     }
     return true;
   };
+
+  bool shouldLower(const Node *N) const override {
+    if (N->getKind() == Kinded::Kind::ConvolutionNodeKind)
```
Comment (opti-mix) [inline lib/Backends/OpenCL/OpenCL.h]: Please add a comment like: "OpenCL backend can handle any kind of convolutions, including group convolutions" or something like that.

File: lib/Backends/OpenCL/OpenCL.cpp [hunk 4]
```diff
@@ -399,14 +399,15 @@ void OpenCLFunction::executeConvolution(const OCLConvolutionInst *CC) {
...
+  unsigned group = CC->getGroup();
   // Create options for compiling the program.
   // Don't use names M, N, K as they are defined in precompiled headers.
 
   std::vector<std::string> options;
   // Number of spacial axes.
   addIntOption(options, "v_nax", 2);
   // Number of groups.
-  addIntOption(options, "v_g", 1);
+  addIntOption(options, "v_g", group);
```
Comment (opti-mix) [inline lib/Backends/OpenCL/OpenCL.cpp]: Add an assert that group is not > 1. And a comment that fast convolution kernels have issues with groups > 1 and that this assert should be removed once these issues are fixed.

Comment (beicy) [reply lib/Backends/OpenCL/OpenCL.cpp]: This assertion has been added at the front of this function.

File: lib/Backends/OpenCL/kernels.cl [hunk 5]
```diff
@@ -755,11 +755,12 @@ __kernel void softmaxgradW(__global void *mem, cl_uint32_t origDest,
 __kernel void convolutionK(__global float *dest, __global float *src,
                            __global float *filter, __global float *bias,
                            ShapeHW filterSizes, ShapeHW strides,
-                           PaddingTLBR pads, ShapeNHWC odim, ShapeNHWC idim,
+                           PaddingTLBR pads, cl_uint32_t group, ShapeNHWC odim, ShapeNHWC idim,
                            ShapeNHWC filterDim) {
   size_t ax = get_global_id(0);
   size_t ay = get_global_id(1);
   size_t d = get_global_id(2);
+  size_t offset = d / (odim.c / group) * (idim.c / group);
```
Comment (artemrakhov) [inline lib/Backends/OpenCL/kernels.cl]: nit: offset is general name, which interferes a bit with scale/offset in Quantization case. How does `inChannelOffset` sound?
Also, let's make code more similar to Interpreter implementation by having `inCperG` and `outCperG`.

File: lib/Backends/OpenCL/kernels.cl [hunk 6]
```diff
@@ -755,11 +755,14 @@ __kernel void softmaxgradW(__global void *mem, cl_uint32_t origDest,
 __kernel void convolutionK(__global float *dest, __global float *src,
                            __global float *filter, __global float *bias,
                            ShapeHW filterSizes, ShapeHW strides,
-                           PaddingTLBR pads, ShapeNHWC odim, ShapeNHWC idim,
+                           PaddingTLBR pads, cl_uint32_t group, ShapeNHWC odim, ShapeNHWC idim,
```
Comment (artemrakhov) [inline lib/Backends/OpenCL/kernels.cl]: nit: Line is longer than 80 chars. Apparently clang-format doesn't work for this file, because it has `.cl` extension. You can temporarily rename it to `.cpp` to run clang-format again. But feel free to do this in follow up PR. There may be other format changes, not related to your PR.
